### ISSUE TITLE
Complex numbers

### ISSUE THREAD

**Issue by rreusser** [dstructs/dstructs#1]
Are complex numbers worthy of a data structure? If so, what should the preferred structure be? A simple object, i.e. `{'real': 1.5, 'imag': -2.0}`? Or would stored inside an object as a typed array be better for enforcing type consistency, i.e. `{'_data': new Float64Array([1.5, -2.0])}`? (With the latter, getters/setters could trivially smooth out the difference, and algorithms could still operate on `_data` directly for the sake of efficiency.)

Followup: I coded up a couple basic algorithms for dealing with complex numbers carefully, e.g. [complex-sqrt](https://github.com/scijs/complex-sqrt/blob/master/index.js). It'd be trivial to copy those into the compute.io world and might be a good way to establish a new convention or two.

**Comment by kgryte** [comment]
Complex numbers are indeed worthy of a data structure. I believe multiple approaches are possible.

1) As you point out, a simple object.

``` javascript
var cmplx = {
  'real': 1.5,
  'imag': -2.0
};
```

2) A two element array.

``` javascript
var cmplx = [ 1.5, -2.0 ];
```

3) A 128 bit buffer. The first 8 bytes would be real, the last 8 would be imaginary. 

``` javascript
var cmplx = new Buffer( 16 );
```

4) Complex typed array flavors.

``` javascript
var cmplx32 = new Float32Array( 2 );
var real = cmplx32[ 0 ];
var imag = cmplx32[ 1 ];
```

Ditto for other dtypes.

5) Within a complex array.

``` javascript
var len = 100;
var carr = new Float64Array( len*2 );
var reals = carr[ 0:2:end ];
var imag = carr[ 1:2:end ];
```

Probably other flavors, as well. One thing to ask when choosing the data structure is what would a matrix of complex numbers look like? If a matrix is typed, then an object/simple array approach is probably not going to work, especially as object arrays are harder for the compiler to optimize.

If we take the approach that a complex number is simply a two element typed array, then the extension to matrices (and beyond) is straightforward. A complex matrix is simply `2*N*M` in size to accommodate the real and imaginary pairs. Getters and setters will be needed, as some indirection (mapping) will be needed to map `(i,j)` to the physical location. 

In thinking about this, this would be an interesting and worthwhile data structure.

**Comment by rreusser** [comment]
Yeah, this is one place where the ndarray solution excels since it's trivial to interleave real/imag elements as a m x n x 2 ndarray and then slice the matrix to get at the real/imag parts. Complex matrix arithmetic is then just marginally suboptimal (because of parallel loop iteration where a simple offset would suffice) map-reduce operations.

Since there's not an infinite profusion of datatypes though, I can see the argument for a cmatrix dstruct that mirrors matrix functionality with the additional assumption that real/imag are strictly interleaved in memory. Instead of a complex dstruct (and to avoid venturing into typed-array-subclassing land), complex numbers are understood to be two-element arrays. (The cmatrix getter could return a two-element [subarray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/subarray) if not for legacy (i.e. IE<10) support.)

More generally speaking, I think there's also room for a numpy-like ufunc wrapper concept so that when someone writes, say, a real (or complex) hyperbolic whatnot function, there's minimal boilerplate.

I know I'm at risk of feature-creep before even starting, but I think a bit of DRYness is a small price to pay to keep things loosely coupled and not sacrifice the long-term possibility of enumerating some of these use-cases by working it into an over-metaprogrammed, tightly coupled state.

**Comment by rreusser** [comment]
On a higher level, my concern here is less about how I'm going to sit down right now and write a bunch of complex number/matrix routines and more about understanding structure/conventions of a loosely-coupled library that doesn't shut out the possibility of ever integrating these concepts in a reasonable manner. In other words, where things fall on the verbose vs DRY-but-unmaintainably-over-metaprogrammed spectrum.
